### DISCUSSION TITLE
State management with dynamic fields and data : looking for feedback

### DISCUSSION THREAD

**Discussion by theredfish** [react-hook-form/react-hook-form#9776] [Q&A] [answered]
Hi,

Since I'm working alone on my project, I'm looking for feedback about my design choice. Btw, I want to thanks the maintainers for the good work. The documentation is really great and so far I built complex form behaviors in no time !

# Example

> react-hook-form:  v7

Here is a [codesandbox](https://codesandbox.io/s/conditional-fields-tc4yoz?file=/src/App.tsx) of the case I currently have (but in a much simpler way) :

- I have a lot of nested components for my form (multiple tabs etc...) so I use `useFormContext`, that's why I'm using it in this example.
- I watch the json object to load selects conditionally 
- I use react-select, the particularity is that I need a `label` and a `value` for the options.
- I have a mix of states
  - the `defaultValues` that I use when I want to provide pre-filled info or when a user edit an existing item
  - the component also provides a state for all the different options.
  - the `data` attached to each option. The particularity is that I don't have the same object that I'm passing. Each option can provide different objects. In the example I provide one select manipulates a `description` attribute where the other manipulates a `list`.

# Issues
1. Major issue so far : I can't "save" the previous state. As you can see, the example app starts with `cheese` + `Roquefort` but if you switch to `wine` and come back to `cheese`, you will see `Camembert`.
2. Other issue : I think this design leads to a bad separation of concerns where the state management is spread into different places as described above (mix of states).

# Feedback
I'm just looking for feedback regarding my design decisions. How can I improve the code? Should I consider to switch to a state manager? Note that I'm new to React (and frontend in general) so stuff like Redux is new for me and I don't see the need yet. I still learn about the frameworks and libraries. In a way, I use RHF as a state manager for my forms since my app is mainly made of forms for the moment.

Thanks!

**Comment by leapful** [comment, marked as answer]
@theredfish 

The problem is not from having state management or not having state management. You are mixing `choice.data` object with multiple types like `cheeseOptions` and `wineOptions`, meaning you are still in mindset storing 2 items into just 1 box that can only store 1 item, therefore, it will be always complex to rollback last selected value of each list option (`cheeseOptions`, `wineOptions`).

The easiest solution is to change your form data structure into separated object data for `cheese` and `wine`, for example:

```
type FormValues = {
   choice: {
      label: string;
      value: string;
      cheese: {
          label: string;
          value: string;
          description: string;
      };
      wine: {
          label: string;
          value: string;
          list: string;
      };
   }

}

// then

{choice === 'wine' && <Controller name="choice.wine">}
{choice === 'cheese' && <Controller name="choice.cheese">}
```

Or if you really needs to keep the old `choice.data` structure, then just use the structure above for your form then transform it into `choice.data` on submit via `handleSubmit` method.

```

const onSubmit = (formValues) => {
   const { wine, cheese, ...newData } = formValues;
   newData.data = newData.value === 'wine' ? wine : cheese;
   console.log(newData);
}

return <form onSubmit={handleSubmit(onSubmit)}>

```

**Reply by theredfish** [reply]
Thank you for your reply. Keeping structures separate makes sense indeed.

My real business case is to store authentication information (ex oauth, basic auth, api key, ...) and I tried to keep common information at the same level and not to repeat myself because I may have many entities to list. But it seems that's not the way RHF works best, so I'll go back to your solution. Also it will be easier to manage types and a solution like `yup` with your approach. Thanks!
